### PR TITLE
5116 bugfix date change on unfinalized steps

### PR DESCRIPTION
**Pull request by emitche**
- Ensure steps that are not finalized can be moved
- Ensure changing a step date with dependents also moves the dependents
- Ensure steps with precedent steps cannot be moved
- Indicate steps with precedent steps on the calendar

### CODE REVIEW

File: app/serializers/calendar_event_serializer.rb [hunk 1]
```diff
@@ -1,8 +1,9 @@
 class CalendarEventSerializer < ActiveModel::Serializer
-  attributes :start, :html, :id, :model_id, :editable, :is_finalized, :completed, :model_type, :event_type
+  attributes :start, :html, :id, :model_id, :editable, :is_finalized, :completed, :model_type, :event_type, :has_precedent
```
Comment (emitche) [inline app/serializers/calendar_event_serializer.rb]: What is the recommended way to list attributes when the line is too long?

Comment (emitche) [inline app/serializers/calendar_event_serializer.rb]: What is the recommended way to list attributes when the line is too long?

File: app/models/calendar_event.rb [hunk 2]
```diff
@@ -94,6 +95,7 @@ def initialize_project_step(step)
     @step_type = step.step_type_value
     @completion_status = step.completion_status
     @time_status = step.days_late && step.days_late > 0 ? "late" : "on_time"
+    @has_precedent = step.schedule_parent_id ? true : false
```
Comment (emitche) [inline app/models/calendar_event.rb]: Do I need true and false here?

File: app/models/calendar_event.rb [hunk 3]
```diff
@@ -14,6 +14,7 @@ class CalendarEvent
   attr_accessor :completion_status
   attr_accessor :time_status
   attr_accessor :num_of_logs
+  attr_accessor :has_precedent
```
Comment (hooverlunch) [inline app/models/calendar_event.rb]: Boolean accessors like this should always have a `?` at the end. So you could either define it as a regular method or you could add `alias_method :has_precedent?, :has_precedent`. I recommend the latter as it's less code and you'll need the non `?` version in the serializer.

File: app/serializers/calendar_event_serializer.rb [hunk 4]
```diff
@@ -1,8 +1,9 @@
...
 
   def editable
     return false if object.event_type == "ghost_step"
+    return false if object.has_precedent
```
Comment (hooverlunch) [inline app/serializers/calendar_event_serializer.rb]: Use the `?` version here. You should also combine these two using `||`.

File: app/views/admin/calendar/_event.html.slim [hunk 5]
```diff
@@ -12,7 +12,8 @@
   - when "project_step"
 
     div.calendar-event.cal-step [
-      class="cal-step-#{cal_event.step_type} #{cal_event.time_status}"
+      class="cal-step-#{cal_event.step_type} #{cal_event.time_status}
+        #{cal_event.has_precedent ? 'has-precedent': ''} "
```
Comment (hooverlunch) [inline app/views/admin/calendar/_event.html.slim]: Very nice. Can use `?` version here too and throughout the templates.

File: app/views/admin/calendar/_legend.html.slim [hunk 6]
```diff
@@ -13,14 +13,23 @@ div#legend-content
             td
             th.bold = t("calendar.status.incomplete")
           tr
+            td: span: i.fa.fa-pencil-square-o
```
Comment (hooverlunch) [inline app/views/admin/calendar/_legend.html.slim]: Nice! I didn't know about this syntax...

File: app/views/admin/calendar/_legend.html.slim [hunk 7]
```diff
@@ -13,14 +13,23 @@ div#legend-content
...
+            th = t("calendar.num_of_logs")
+          tr
+            td: span.original-date
+            th = t("calendar.original_date")
+          tr
+            td: span: i.fa.fa-link
+            th = t("calendar.has_precedent")
```
Comment (hooverlunch) [inline app/views/admin/calendar/_legend.html.slim]: Very nice to remember this!

File: config/routes.rb [hunk 8]
```diff
@@ -46,7 +46,11 @@
     resources :project_groups
 
     # Does it make sense to surround in separate namespace?
-    resources :timeline_step_moves
+    resources :timeline_step_moves do
```
Comment (hooverlunch) [inline config/routes.rb]: Can we get rid of the comment on line 48 please.

File: app/controllers/admin/timeline_step_moves_controller.rb [hunk 9]
```diff
@@ -27,6 +27,14 @@ def create
     render nothing: true
   end
 
+  # A change of date for a step that does not require a corresponding log
```
Comment (emitche) [inline app/controllers/admin/timeline_step_moves_controller.rb]: @hooverlunch: Is this enough for a message? I spoke with linesarefuzzy about adding this method to this file specifically. There's a bit more going on here than just the saving of the log. The other methods refer to the Timeline StepMove model, but that's not needed here. We are simply changing a date in this method.

**Pull request merged**